app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/CheckAllAnswersAndSubmitController.scala

```diff
@@ -72,33 +73,63 @@ class CheckAllAnswersAndSubmitController @Inject() (
           )
         val result =
           for {
-            response <- returnsService.submitReturn(submittedReturnRequest)
+            response <- returnsService
+                         .submitReturn(submittedReturnRequest)
+                         .map(Right(_))
+                         .leftFlatMap[Either[Error, SubmitReturnResponse], Error](e => EitherT.pure(Left(e)))
```

Comment:
I don't see the point of the `leftFlatMap` here. What have you achieved?

Reply:
The error is now on the right of the either of the EitherT so I can change the journey status just below it .in the for-comprehension. It's not pretty, but the type of this part of the for-comprehension is `EitherT[Future, Error, Either[Error, SubmitReturnResponse]]`

Reply:
Yeah, it's not good. I see why you have done it - you're trying to capture two different possible errors from two different effects in one for-comprehension. The problem - I think - boils down to the error channel in EitherT not being covariant. Because it is not and the fact that Error is final you cannot build a coproduct to reflect the different type of errors in the service e.g SessionUpdateError, SubmissionError, etc etc...

Reply:
In this case it's more that you can't chain on the error type in the for comprehension with EitherT since flatMap always maps on the right type. That's why it looks like that - I wanted to be able to update the session on the for comprehension in the submit error case.

Reply:
So even if I was able to build a coproduct it still won't allow me to map on the left in the EitherT in the for comprehension

Reply:
Do you think by changing the success type of `returnsService.submitReturn` to return the correct type you need i.e SubmitRetunFailed or JustSubmittedReturn and then case matching would be clearer?